internal/filters.go

```diff
@@ -0,0 +1,208 @@
+func (mf *SplitFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	// always splits into two identical events
+	events := make([]*Event, 0)
+	//TODO: implement filter logic
+	//TODO: clone
+	events = append(events, event)
+	events = append(events, event)
+	return events, nil
+}
+
+// Filter lets any event pass
+func (mf *PassFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	return []*Event{event}, nil
+}
+
+// Filter lets no event pass
+func (mf *BlockFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	return []*Event{}, nil
+}
+
+// NewFilterer is a factory function to create appropriate filterer for given filter plugin config
+func NewFilterer(ctx context.Context, fp *FilterPlugin) (Filterer, error) {
+	if fp == nil {
+		return nil, new(MissingFilterPluginConfigError)
+	}
+	switch fp.Type {
+	case FilterTypeFilter:
+		flt := new(FilterFilter)
+		flt.Matcher = NewDefaultPatternMatcher()
+		// parse config params
+		if fp.Params != nil {
+			if value, ok := fp.Params["pattern"]; ok {
+				flt.Pattern = value
+			}
+		}
+		return flt, nil
+	case FilterTypeMatcher:
+		flt := new(MatchFilter)
+		flt.Matcher = NewDefaultPatternMatcher()
+		// parse config params
+		if fp.Params != nil {
+			if value, ok := fp.Params["pattern"]; ok {
+				flt.Pattern = value
+			}
+		}
+		return flt, nil
+	case FilterTypeTransformer:
+		//TODO: pass in config params here
+		return new(TransformFilter), nil
+	case FilterTypeTTLer:
+		//TODO: pass in config params here
+		return new(TTLFilter), nil
+	case FilterTypeSplitter:
+		//TODO: pass in config params here
+		return new(SplitFilter), nil
+	case FilterTypePass:
+		return new(PassFilter), nil
+	case FilterTypeBlock:
+		return new(BlockFilter), nil
+	}
+	return nil, &UnknownFilterTypeError{fp.Type}
+}
+
+// DoSync synchronoulsy accepts an event into a filter plugin belonging to a filter chain
+func (fp *FilterPlugin) DoSync(ctx context.Context, event *Event) error {
+	log.Debug().Msg(fp.Type + " filter " + fp.Hash(ctx) + " passed")
+	fp.EventCount++
+	filteredEvents, err := fp.filterer.Filter(ctx, event)
+	if err != nil {
+		return err
+	}
+	for _, e := range filteredEvents {
+		if fp.outputChannel != nil {
+			fp.outputChannel <- e
+		}
+	}
+	return nil
+}
+
+// DoAsync kicks of a go routine listening indefinitly for a events to arrive on its in channel
+func (fp *FilterPlugin) DoAsync(ctx context.Context) {
+	go func() {
+		if fp.inputChannel == nil {
+			return
+		}
+		for {
+			var inputEvent *Event
+			select {
+			case inputEvent = <-fp.inputChannel:
+				fp.EventCount++
+			case <-fp.done:
+				log.Debug().Msg(fp.Mode + " " + fp.Type + " plugin " + fp.Hash(ctx) + " done")
+				return
+			}
+			filteredEvents, err := fp.filterer.Filter(ctx, inputEvent)
+			if err != nil {
+				return
+			}
+			if len(filteredEvents) > 0 {
+				log.Debug().Msg(fp.Mode + " " + fp.Type + " plugin " + fp.Hash(ctx) + " passed")
```

Comment:
Please log with log.Ctx(ctx).Debug()

Reply:
Also, I will stop calling them out, but please go through the logs, and update all of them with log.Ctx(ctx). Thanks!